src/neo/SmartContract/InteropService.Oracle.cs

```diff
@@ -0,0 +1,36 @@
+using Neo.Oracle.Protocols.HTTP;
+using System;
+
+namespace Neo.SmartContract
+{
+    partial class InteropService
+    {
+        public static class Oracle
+        {
+            public static readonly uint System_Oracle_Get = Register("System.Oracle.Get", Oracle_Get, 0, TriggerType.Application, CallFlags.None);
+
+            private static bool Oracle_Get(ApplicationEngine engine)
+            {
+                if (engine.OracleCache == null) return false;
+                if (!engine.TryPop(out string urlItem) || !Uri.TryCreate(urlItem, UriKind.Absolute, out var url)) return false;
+                if (!engine.TryPop(out string filter)) return false;
+
+                var request = new OracleHTTPRequest()
+                {
+                    Method = OracleHTTPRequest.HTTPMethod.GET,
+                    URL = url,
+                    Filter = filter,
+                    Body = null
+                };
+
+                if (engine.OracleCache.TryGet(request, out var response))
```

Comment:
In what situation will this `TryGet` return `false`?

Reply:
If the oracle is the caller, never. Because he will download and return the respose (or the error).
If is an oracle TX and it's already cached, it has a chance of return false if it was not found in the cache. For example:

`Oracle.Get("https://myapi.com?index="+Blockchain.Height);`

With this request, the oracle can process one request, cache it, and when the persist process this TX, it's possible that the request changed, so it can't be possible to find it in the cache, and need to return FAULT.

Reply:
I think we should read the uri from the attributes. This ensures the invariance of the uri.

Reply:
The hash of the request is done with all the uri, so if it change, the result is not in the cache.

>This ensures the invariance of the uri.

How do you want to ensure this?

Reply:
Different oracle nodes process the oracle tx, may create different `OracleRequest`. For example, one oracle node request once, another request twice, and it may also be a valid oracle tx packaged in block with one request.

Reply:
@shargon How do you read the URIs before executing the contract? The only way is to put the URIs into the attributes. In contract, we do not need to put the URIs again. We can read them from the attributes directly.

Reply:
I think that is not a problem, if there different request, the consensus is like different responses, it's expected

Reply:
At the end, you will have a different cache. You can see an example here

https://github.com/neo-project/neo/pull/1243/files#diff-bb673ce47bba9d855fbfbb9afbe90c99R69

Reply:
@shargon How do you read the URIs before executing the contract?

Reply:
Now it's not possible to read before execute the contract. The request is done while the execution, only if it is a Oracle node

Reply:
So the contract will be executed twice? I don't think it's a good idea.

Reply:
We will read it first in order to know if it has the oracle attribute (if is an oracle tx).
if it's an oracle tx, then we will wait until get the OracleResult. This is the result of the execution /consensus of the oracle nodes.

It's executed only once by nodes, because they wait for the oracle result, the oracle nodes will execute twice, but the second has no internet connections, because it's already cached.

Reply:
The results of the two executions may be inconsistent. We must avoid this situation.

Reply:
It could be, but I think that it's not a problem, because only one is for download, when we have this information, it will be used `onPersist` for the execution. The real result is this, if the second execution try to download something different, it will FAULT.

![image](https://user-images.githubusercontent.com/3167973/72448128-cfe9c100-37b6-11ea-8f0f-8e97abdb7554.png)


What it's your idea?

Reply:
If the first execution is `FAULT`, it won't download anything. But the second one could be `HALT`.

Reply:
The first execution is executed in the node who create the TX, in order to create the Attributes, so if it's FAULT, the tx should not be relayed.

Then if the Oracle decide that it's a FAULT, they will stamp the cache with the result that must return FAULT (if the download requests are exactly this), here we have 3 scenarios.

1. The execution during `OnPersist` has the same request as the oracles. OK, take all the information from the cache, and use it.

2. There are some request diferent from the Oracle `http://google.es?index={Blockchain.Index}` then the TX will FAULT, because the information is not stored.

3. Now the download cache it's not used because of a condition `if (height > 1000) download("google.es")`. HALT or FAULT according to the execution, but there are no problem, because the execution is without oracle, and the oracle was paid.

Reply:
> `http://google.es?index={Blockchain.Index}`

I don't know why we should give it a chance to create non-deterministic URIs. We should only allow immutable URIs.

Reply:
But how can we do that? we need to be able to check dynamic information

`https://myserver.com/apiuserInfo?=1234142`

Reply:
Then I prefer to remove the `TransactionAttribute`. It is never used.

Reply:
It's used for store the expected hash (if users want to) and know if the transaction it's an oracle transaction

Reply:
If the URI is dynamic, how is it possible to know the expected hash in advance?

Reply:
> If the URI is dynamic, how is it possible to know the expected hash in advance?

 He can choose not to set the expected hash.

Reply:
If so, why do we need it?

Reply:
We have discussed with @belane  before at https://github.com/neo-project/neo/issues/1274#issuecomment-559860787

Reply:
All the execution could be executed before in the node, to establish what result is expected, because you may want to buy rice, but only if the API returns its price, not another.